papermill/engines.py

```diff
@@ -0,0 +1,328 @@
+        Optionally called by engines during execution to initialize the
+        metadata for a cell and save the notebook to the output path.
+        """
+        cell.metadata.papermill['start_time'] = self.now().isoformat()
+        cell.metadata.papermill["status"] = self.RUNNING
+        cell.metadata.papermill['exception'] = False
+
+        self.save()
+
+    @catch_nb_assignment
+    def cell_exception(self, cell, **kwargs):
+        """
+        Called by engines when an exception is raised within a notebook to
+        set the metadata on the notebook indicating the location of the
+        failure.
+        """
+        cell.metadata.papermill['exception'] = True
+        cell.metadata.papermill['status'] = self.FAILED
+        self.nb.metadata.papermill['exception'] = True
+
+    @catch_nb_assignment
+    def cell_complete(self, cell, **kwargs):
+        """
+        Optionally called by engines during execution to finalize the
+        metadata for a cell and save the notebook to the output path.
+        """
+        end_time = self.now()
+        cell.metadata.papermill['end_time'] = end_time.isoformat()
+        if cell.metadata.papermill.get('start_time'):
+            start_time = dateutil.parser.parse(cell.metadata.papermill['start_time'])
+            cell.metadata.papermill['duration'] = (end_time - start_time).total_seconds()
+        if cell.metadata.papermill['status'] != self.FAILED:
+            cell.metadata.papermill['status'] = self.COMPLETED
+
+        self.save()
+        if self.pbar:
+            self.pbar.update(1)
+
+    @catch_nb_assignment
+    def notebook_complete(self, **kwargs):
+        """
+        Finalizes the metadata for the notebook and saves the notebook to
+        the output path.
+
+        Called by EngineBase when execution concludes, regardless of exceptions.
+        """
+        self.end_time = self.now()
+        self.nb.metadata.papermill['end_time'] = self.end_time.isoformat()
+        if self.nb.metadata.papermill.get('start_time'):
+            self.nb.metadata.papermill['duration'] = (
+                self.end_time - self.start_time
+            ).total_seconds()
+
+        # Cleanup cell statuses in case callbacks were never called
+        for cell in self.nb.cells:
+            if cell.metadata.papermill['status'] == self.FAILED:
+                break
+            elif cell.metadata.papermill['status'] == self.PENDING:
+                cell.metadata.papermill['status'] = self.COMPLETED
+
+        self.complete_pbar()
+        self.cleanup_pbar()
+
+        # Force a final sync
+        self.save()
+
+    def complete_pbar(self):
+        if hasattr(self, 'pbar') and self.pbar:
+            self.pbar.n = len(self.nb.cells)
+            self.pbar.refresh()
+
+    def cleanup_pbar(self):
+        if hasattr(self, 'pbar') and self.pbar:
+            self.pbar.close()
+            self.pbar = None
+
+    def __del__(self):
+        self.cleanup_pbar()
+
+
+class EngineBase(object):
```

Comment:
Calibration question: `Engine` seems like a much nicer name than `EngineBase`, am I the only one who is bothered by this? I can't rationally explain why it bothers me, it just does, like a picture that isn't hanging perfectly straight ... :)

Reply:
I've typically used `Base` as suffix for abstract classes. I'm fine with changing it to `Engine` though.

Reply:
This isn't an abstract class though, since it's inheriting from object. If you were to want to make it an actual abstract class, one solution I've liked elsewhere in the Jupyter ecosystem is appending `ABC` to the end of the name.